stellar-contract-env-host/src/host.rs

```diff
@@ -177,6 +177,27 @@ struct RollbackPoint {
     objects: usize,
 }
 
+#[cfg(feature = "testutils")]
+pub trait ContractVTableFnTrait {
+    fn call(&self, args: &[RawVal]) -> RawVal;
+
+    fn duplicate(&self) -> ContractVTableFn;
```

Comment:
What is `duplicate`'s behavior, and how is it different to the behavior of `clone`?

Reply:
Actually, why do you need to be able to clone fns? That seems surprising.

Reply:
This goes a couple of layers deep. The root cause is that `EnvBase` supports `deep_clone`, which facilitates step-forward/step-backward state checking. For `Host`, this is implemented in terms of `HostImpl::clone`. Therefore `HostImpl` is `Clone`, which means that all parts of `HostImpl` need to be `Clone` as well. Specifically, `RefCell<std::collections::HashMap<Hash, ContractVTable>>` needs to be `Clone`, which is the case only if `ContractVTable` is `Clone`, which is the case only if `ContractVTableFn` is `Clone`. In summary, if we want to change this, then we need to change how step-forward/step-backward state checking works.

So how is `duplicate` different from `clone`? `Box<dyn Clone>` is not [object safe](https://doc.rust-lang.org/reference/items/traits.html#object-safety) because `Clone` uses `Self` outside of the receiver. This `duplicate` machinery, in contrast, is object safe. It works in conjunction with `Box<...>` by constructing a new `Box<...>` by cloning the boxed object.

Reply:
It might be possible to provide a custom implementation of `Clone` for `std::collections::HashMap<Hash, ContractVTable>` using the newtype idiom. I don't know if that will be better, but I will investigate now.

Reply:
It took a bit of finagling, but I think I was able to get this cleaned up. Is it better? Maybe. Did it allow me to find a different design flaw and fix it? Yes.

We can do more work on this interface in the future, but I don't want to block this PR on getting the best possible interface right now. It's important to remember that my goal is for this machinery to end up behind a macro. I'm already partially doing that in rs-stellar-token-contract.